src/instance.rs

```diff
@@ -32,21 +32,142 @@ pub unsafe trait PyNativeType: Sized {
+///         // Foo is no longer valid.
+///         // Returning it from this function is a 💥 compiler error 💥
+///         foo
+///     }
+/// }
+/// ```
+///
+/// [`Py`]`<T>` can be used to get around this by converting `dict` into a GIL-independent reference:
+///
+/// ```rust
+/// # use pyo3::prelude::*;
+/// # use pyo3::types::PyDict;
+/// #
+/// struct Foo {
+///     inner: Py<PyDict>,
+/// }
+///
+/// impl Foo {
+///     fn new() -> Foo {
+///         Python::with_gil(|py| {
+///             let dict: Py<PyDict> = PyDict::new(py).into();
+///             Foo { inner: dict }
+///         })
+///     }
+/// }
+/// ```
+///
+/// This can also be done with other pyclasses:
+/// ```rust
+/// # use pyo3::prelude::*;
+/// #
+/// #[pyclass]
+/// struct Bar {/* fields omitted */}
+///
+/// struct Foo {
+///     inner: Py<Bar>,
+/// }
+///
+/// impl Foo {
+///     fn new() -> PyResult<Foo> {
+///         Python::with_gil(|py| {
+///             let bar: Py<Bar> = Py::new(py, Bar {})?;
+///             Ok(Foo { inner: bar })
+///         })
+///     }
+/// }
+/// ```
+///
+/// # Example: Shared ownership of Python objects
+///
+/// `Py<T>` can be used to share ownership of a Python object, similar to std's [`Rc`]`<T>`.
+/// As with [`Rc`]`<T>`, cloning it increases its reference count rather than duplicating
+/// the underlying object.
+///
+/// This can be done using either [`Py::clone_ref`] or [`Py`]`<T>`'s [`Clone`] trait implementation.
+/// [`Py::clone_ref`] will be faster if you happen to be already holding the GIL.
+///
+/// ```rust
+/// use pyo3::prelude::*;
+/// use pyo3::types::PyDict;
+/// use pyo3::conversion::AsPyPointer;
+///
+/// # fn main() {
+/// Python::with_gil(|py| {
+///     let first: Py<PyDict> = PyDict::new(py).into();
+///
+///     // All of these are valid syntax
+///     let second = Py::clone_ref(&first, py);
+///     let third = first.clone_ref(py);
+///     let fourth = Py::clone(&first);
+///     let fifth = first.clone();
+///
+///     // Disposing of our original `Py<PyDict>` just decrements the reference count.
+///     drop(first);
+///
+///     // They all point to the same object
+///     assert_eq!(second.as_ptr(), third.as_ptr());
+///     assert_eq!(fourth.as_ptr(), fifth.as_ptr());
+///     assert_eq!(second.as_ptr(), fourth.as_ptr());
+/// });
+/// # }
+/// ```
+///
+/// # Preventing reference cycles
+///
+/// It is easy to accidentally create reference cycles using [`Py`]`<T>`.
+/// The Python interpreter can break these reference cycles within pyclasses if they
+/// implement the [`PyGCProtocol`](crate::class::gc::PyGCProtocol). If you permit your pyclass
+/// to create reference cycles you should implement this protocol to avoid leaking memory.
+///
+/// # A note on `Send` and `Sync`
+///
+/// Accessing this object is threadsafe, since any access to its API requires a [`Python<'py>`](crate::Python) token.
+/// As you can only get this by acquiring the GIL, `Py<...>` is safe against concurrent access.
```

Comment:
replace "safe against concurrent access" which is imprecise with "implements Send and Sync".